### PR TITLE
Document the load balancer feature

### PR DESCRIPTION
**Pull request by mangelajo**
Signed-off-by: Miguel Angel Ajo <majopela@redhat.com>

### CODE REVIEW

**Review by nyechiel — COMMENTED** [review]
Thanks for working on this, @mangelajo. Other than a new user guide, anything we should mention here https://submariner.io/getting-started/#prerequisites? I also think it's a good idea to touch on this feature and key unsupported/untested combinations under https://submariner.io/operations/known-issues/

**Comment by mangelajo** [comment]
Thanks @nyechiel I updated it with your comments, as for https://submariner.io/getting-started/#prerequisites I'm still not sure how to update it. I guess on the part where we say that a public IP is necessary for the gateways.

File: src/content/getting-started/quickstart/openshift/aws-lb/_index.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+title: "On AWS (LoadBalancer mode)"
+weight: 20
+---
+
+This quickstart guide covers the necessary steps to deploy two OpenShift Container Platform (OCP) clusters on AWS with
+a cloud network load balancer service in front of the gateways and without dedicated gateways.
+
+In this mode no dedicated gateways are necessary. The administrator will need to manually label some
+nodes as Submariner gateways, and the submariner-operator will take care of creating a LoadBalancer
+type of Service pointing to the active submariner-gateways.
```
Comment (nyechiel) [inline src/content/getting-started/quickstart/openshift/aws-lb/_index.md]:
```suggestion
This quickstart guide covers the necessary steps to deploy two OpenShift Container Platform (OCP) clusters on AWS leveraging
a cloud network load balancer service in front of the Submariner gateways.

The main benefit of this mode is that there is no need to dedicate specialized nodes with a public IP address to act as gateways. The administrator only needs to manually label any existing
node or nodes in each cluster as Submariner gateways, and the submariner-operator will take care of creating a LoadBalancer
type Service pointing to the active Submariner gateway.
```

**Comment by nyechiel** [comment]
> Thanks @nyechiel I updated it with your comments, as for https://submariner.io/getting-started/#prerequisites I'm still not sure how to update it. I guess on the part where we say that a public IP is necessary for the gateways.

Yes, that's what I was thinking of. I think that other sections (for e.g the architecture part) will need to be updated, but it makes sense to change the minimum required now, and improve with 0.11 once we have more testing and experience with this new mode.

File: src/content/getting-started/quickstart/openshift/aws-lb/_index.md (hunk 1 above)
Comment (mangelajo) [reply src/content/getting-started/quickstart/openshift/aws-lb/_index.md]: Done :)

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 2]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+
+{{% notice info %}}
+The load balancer mode is still experimental, we need to test in different clouds and characterize
+failover times.
+{{% /notice %}}
+
+The load balancer mode is designed to simplify the deployment of Submariner in cloud deployments
+where worker nodes don't have a dedicated public IP available.
+
+When enabled for a cluster during join [`subctl`](../../../operations/deployment/subctl/#join-flags-general),
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
When enabled for a cluster during [`subctl join`](../../../operations/deployment/subctl/#join-flags-general),
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 3]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+the operator will create a LoadBalancer type Service exposing both the encapsulation dataplane port
+as well as the natt discovery port. This load balancer targets pods labeled with
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
as well as the NAT-T discovery port. This load balancer targets pods labeled with
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 4]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+`gateway.submariner.io/status=active` and `app=submariner-gateway`
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
`gateway.submariner.io/status=active` and `app=submariner-gateway`.
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 5]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+
+When the LoadBalancer mode is enabled the preferred server mode is enabled automatically for
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
When the LoadBalancer mode is enabled, the `preferred-server` mode is enabled automatically for
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 6]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+the cluster, as IPSEC is incompatible with the bi-directional connection mode and the
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
the cluster, as IPsec is incompatible with the bi-directional connection mode and the
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 7]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+load balancers and client/server connectivity is necessary.
+
+![Figure 3 - Gateway behind load balancer](/images/high-availability/HA_Cluster_LB1.png)
+
+If a failover occurred, the load balancer would update to the new available and active
+gateway endpoints.
+
+![Figure 4 - Gateway behind load balancer failover](/images/high-availability/HA_Cluster_LB2.png)
+
+### Preferred server mode
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
### Preferred-server mode
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 8]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+
+This mode modifies the behaviour of IPSEC based cable drivers, other cable drivers ignore
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
This mode is only specific to the libreswan cable-driver which is based on IPsec. Other cable drivers ignore
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 9]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+this setting.
+
+When enabled for a cluster during join [`subctl`](../../../operations/deployment/subctl/#join-flags-general),
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
When enabled for a cluster during [`subctl join`](../../../operations/deployment/subctl/#join-flags-general),
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 10]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+the gateway will try to establish connection with other clusters by configuring the IPSEC connection
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
the gateway will try to establish connection with other clusters by configuring the IPsec connection
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 11]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+in server mode, and waiting for remote connections.
+
+Remote clusters will identify the preferred-server mode of this cluster, and attempt the connection.
+This is useful in environments where on-premises clusters don't have access to port mapping.
+
+When both sides of a connection are in preferred server mode, they will compare the endpoint cable
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
When both sides of a connection are in `preferred-server` mode, they will compare the endpoint cable
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 12]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+### Preferred server mode
+
+This mode modifies the behaviour of IPSEC based cable drivers, other cable drivers ignore
+this setting.
+
+When enabled for a cluster during join [`subctl`](../../../operations/deployment/subctl/#join-flags-general),
+the gateway will try to establish connection with other clusters by configuring the IPSEC connection
+in server mode, and waiting for remote connections.
+
+Remote clusters will identify the preferred-server mode of this cluster, and attempt the connection.
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
Remote clusters will identify the `preferred-server` mode of this cluster, and attempt the connection.
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 13]
```diff
@@ -73,3 +73,45 @@ to avoid unhealthy clusters during Service discovery.
...
+This is useful in environments where on-premises clusters don't have access to port mapping.
+
+When both sides of a connection are in preferred server mode, they will compare the endpoint cable
+names to decide which one will be server, and which one will be client. When cable names are ordered
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
names to decide which one will be server and which one will be client. When cable names are ordered
```

File: src/content/operations/deployment/subctl/_index.en.md [hunk 14]
```diff
@@ -98,6 +98,9 @@ deployment.
 | `--clustercidr` `<string>`         | Specifies the cluster's CIDR used to generate Pod IP addresses. If not specified, `subctl` will try to discover it and if unable to do so, it will prompt the user
 | `--label-gateway`                  | Label getways (enabled by default). `--label-gateway=false` disables the prompt for a Worker node to use as gateway
 | `--pod-debug`                      | Enable Submariner pod debugging (verbose logging in the deployed pods)
+| `--load-balancer`                  | Enable a cloud loadbalancer in front of the gateways. This removes the need for dedicated nodes with a public IP address.
+| `--preferred-server`               | Enable this cluster as a preferred server for dataplane connections
```
Comment (nyechiel) [inline src/content/operations/deployment/subctl/_index.en.md]:
```suggestion
| `--preferred-server`               | Enable this cluster as a preferred IPsec server for dataplane connections (only available with libreswan cable driver) 
```

**Review by nyechiel — COMMENTED** [review]
Few nits, thanks @mangelajo

**Comment by nyechiel** [comment]
Oh @mangelajo the Markdown linter is complaining

**Comment by mangelajo** [comment]
Thanks for all the comments & fixes, I pushed a new version with the comments applied, and linter issues handled.

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 15]
```diff
@@ -74,3 +74,45 @@ to avoid unhealthy clusters during Service discovery.
...
 [`subctl`](../../../operations/deployment/subctl/#join-flags-healthcheck) join command.
+
+### Load Balancer mode
+
+{{% notice info %}}
+The load balancer mode is still experimental, we need to test in different clouds and characterize
+failover times.
+{{% /notice %}}
+
+The load balancer mode is designed to simplify the deployment of Submariner in cloud deployments
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
The load balancer mode is designed to simplify the deployment of Submariner in cloud environments
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 16]
```diff
@@ -74,3 +74,45 @@ to avoid unhealthy clusters during Service discovery.
...
+where worker nodes don't have a dedicated public IP available.
+
+When enabled for a cluster during [`subctl join`](../../../operations/deployment/subctl/#join-flags-general),
+the operator will create a LoadBalancer type Service exposing both the encapsulation dataplane port
+as well as the NAT-T discovery port. This load balancer targets pods labeled with
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
as well as the NAT-T discovery port. This load balancer targets Pods labeled with
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 17]
```diff
@@ -74,3 +74,45 @@ to avoid unhealthy clusters during Service discovery.
...
+![Figure 3 - Gateway behind load balancer](/images/high-availability/HA_Cluster_LB1.png)
+
+If a failover occurred, the load balancer would update to the new available and active
+gateway endpoints.
+
+![Figure 4 - Gateway behind load balancer failover](/images/high-availability/HA_Cluster_LB2.png)
+
+### Preferred-server mode
+
+This mode is only specific to the libreswan cable-driver which is based on IPsec. Other cable drivers ignore
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
This mode is specific to the libreswan cable-driver which is based on IPsec. Other cable drivers ignore
```

File: src/content/operations/deployment/subctl/_index.en.md [hunk 18]
```diff
@@ -98,6 +98,9 @@ deployment.
 | `--clustercidr` `<string>`         | Specifies the cluster's CIDR used to generate Pod IP addresses. If not specified, `subctl` will try to discover it and if unable to do so, it will prompt the user
 | `--label-gateway`                  | Label getways (enabled by default). `--label-gateway=false` disables the prompt for a Worker node to use as gateway
 | `--pod-debug`                      | Enable Submariner pod debugging (verbose logging in the deployed pods)
+| `--load-balancer`                  | Enable a cloud loadbalancer in front of the gateways. This removes the need for dedicated nodes with a public IP address.
```
Comment (nyechiel) [inline src/content/operations/deployment/subctl/_index.en.md]:
```suggestion
| `--load-balancer`                  | Enable a cloud loadbalancer in front of the gateways. This removes the need for dedicated nodes with a public IP address
```

File: src/content/getting-started/architecture/gateway-engine/_index.en.md [hunk 19]
```diff
@@ -74,3 +74,45 @@ to avoid unhealthy clusters during Service discovery.
 
 The health checking feature can be enabled/disabled via an option on the
 [`subctl`](../../../operations/deployment/subctl/#join-flags-healthcheck) join command.
+
+### Load Balancer mode
+
+{{% notice info %}}
+The load balancer mode is still experimental, we need to test in different clouds and characterize
+failover times.
```
Comment (nyechiel) [inline src/content/getting-started/architecture/gateway-engine/_index.en.md]:
```suggestion
The load balancer mode is still experimental, and is yet to be tested in all cloud providers nor in different failover scenarios.
```

**Review by nyechiel — APPROVED** [review]

**Comment by mangelajo** [comment]
:rocket:

**Review by skitt — APPROVED** [review]

**Pull request merged**